server/src/main/java/org/elasticsearch/search/vectors/KnnVectorQueryBuilder.java

```diff
@@ -23,56 +24,131 @@
 import org.elasticsearch.index.query.QueryBuilder;
 import org.elasticsearch.index.query.QueryRewriteContext;
 import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ObjectParser;
+import org.elasticsearch.xcontent.ParseField;
 import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
 
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 import java.util.Objects;
+import java.util.function.Supplier;
+
+import static org.elasticsearch.common.Strings.format;
+import static org.elasticsearch.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.xcontent.ConstructingObjectParser.optionalConstructorArg;
 
 /**
  * A query that performs kNN search using Lucene's {@link org.apache.lucene.search.KnnFloatVectorQuery} or
  * {@link org.apache.lucene.search.KnnByteVectorQuery}.
  *
- * NOTE: this is an internal class and should not be used outside of core Elasticsearch code.
  */
 public class KnnVectorQueryBuilder extends AbstractQueryBuilder<KnnVectorQueryBuilder> {
     public static final String NAME = "knn";
+    private static final int NUM_CANDS_LIMIT = 10000;
+    public static final ParseField FIELD_FIELD = new ParseField("field");
+    public static final ParseField NUM_CANDS_FIELD = new ParseField("num_candidates");
+    public static final ParseField QUERY_VECTOR_FIELD = new ParseField("query_vector");
+    public static final ParseField QUERY_VECTOR_BUILDER_FIELD = new ParseField("query_vector_builder");
+    public static final ParseField VECTOR_SIMILARITY_FIELD = new ParseField("similarity");
```

Comment:
I am not 100% sure we should have `similarity` here in the `knn` query object.

The reason it was added to the top level `knn` object was because we couldn't have a `vectorSimilarity` query. But now we have opened the flood gates for allowing more than one kind of vector query.

Do we think we should have `similarity` here or not?

Reply:
@benwtrent Thanks for the comment. 
In my opinion, it seems excessive to introduce a new `vectorSimilarity` query just to limit scores of a knn query.
I think we either should:
1. Remove `similarity` from a knn query, and ask users to use `min_score` search request parameter. But, the user should provide minimum score (not similarity)
2. Keep `similarity` parameter in the query. 

WDYT of these options?

Reply:
`min_score` in the search request applies to the entire search clause. It seems like it misses a key usage where `knn` is used within a `filter` clause and executed with aggregations. I don't think we can use the search request `min_score` as a substitute.

I also see that `function_score` has a `min_score` parameter :(. But that thing is so complicated.

So, my counter options would be:
 - keep it in the query
 - we remove it and add a separate one called `vectorSimilarity`.

I am just trying to think of what we would have done if `knn` was always a query and we added similarity thresholds later. Would we have made it a separate query or added a parameter?

Reply:
I agree that `min_score` doesn't seem like a good candidate since it's global within a search. If we continue down the path of additional queries to simplify the API then it makes sense to me to have a wrapper for either a min_score query in general or a specific threshold query to wrap a knn query.

Reply:
I am ok with having an extra `vector_similarity` query that wraps a knn query or `semantic_search` query (a new query that we will introduce for query_vector_builder.

```js
  "vector_similarity": {
    "knn": {
	    "field": "dense-vector-field",
	    "num_candidates": 100,
	    "query_vector" : [...],
	    "text": "The opposite of blue"
  	},
  	"similarity" : 10
  }
```

```js
  "vector_similarity": {
    "semantic_search": {
	    "field": "dense-vector-field",
	    "num_candidates": 100,
	    "model_id": "my-text-embedding-model",
	    "text": "The opposite of blue"
  	},
  	"similarity" : 10
  }
```

Reply:
@giladgal What would be your opinion?

Reply:
I don't think there is an obvious correct choice between these two options. Having a parent query seems more organized and then that same query could be used with different types of vector queries, as long as you know which queries it can be used with. 

Personally I prefer a parameter than a parent query. I think having all vectors being candidates for any query vector is not how we humans think of similarity and relevance. When the engine returns irrelevant results just because they are the most similar results, it is probably not what the end user wanted. In other words I see the similarity threshold as a natural part of a vector search and I would like for it to be documented so that anyone that reads about vector search immediately becomes aware of that important option. I also like that a parameter is simpler to add and I find it easier to read.